Google.Ads.Gax/src/Util/ExceptionUtilities.cs

```diff
@@ -0,0 +1,197 @@
+﻿// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Ads.Gax.Lib;
+using Google.Apis.Auth.OAuth2.Responses;
+using Grpc.Core;
+using System;
+using System.Linq;
+using System.Reflection;
+
+namespace Google.Ads.Gax.Util
+{
+    /// <summary>
+    /// Extensions to <see cref="RpcException"/> class to extract OAuth errors.
+    /// </summary>
+    public static class ExceptionUtilities
+    {
+        /// <summary>
+        /// Parses the OAuth exceptions.
+        /// </summary>
+        /// <param name="e">The aggregate exception that wraps the rpc exception.</param>
+        /// <returns>The parsed exception, or <code>null</code> if the exception cannot be parsed.
+        /// </returns>
+        internal static TokenResponseException ParseOAuthException(AggregateException e) =>
+            ParseOAuthException(ExtractRpcException(e));
+
+        /// <summary>
+        /// Parses the OAuth exceptions.
+        /// </summary>
+        /// <param name="rpcException">The rpc exception</param>
+        /// <returns>The parsed exception, or <code>null</code> if the exception cannot be parsed.
+        /// </returns>
+        internal static TokenResponseException ParseOAuthException(RpcException rpcException)
+        {
+            const string EXCEPTION_TYPE_PREFIX =
+                "Google.Apis.Auth.OAuth2.Responses.TokenResponseException: ";
+            const string ERROR_PREFIX = "Error:";
+            const string DESC_PREFIX = "Description:";
+            const string URI_PREFIX = "Uri:";
+
+            if (rpcException == null)
+            {
+                return null;
+            }
+            if (!(rpcException.StatusCode == StatusCode.Internal ||
+                rpcException.StatusCode == StatusCode.Unavailable))
+            {
+                return null;
+            }
+            int prefixIndex = rpcException.Message.IndexOf(EXCEPTION_TYPE_PREFIX);
+            if (prefixIndex == -1)
```

Comment:
What does -1 mean? I'm assuming that you're trying to catch that there isn't anything in the message. If so, maybe a small comment should help make the magic number more clear.

Reply:
Done.